docs/fsharp/style-guide/conventions.md

```diff
@@ -0,0 +1,648 @@
+
+let closureTableContains (key, value) (t: Dictionary<_, HashSet<_>>) =
+    match t.TryGetValue(key) with
+    | (true, v) -> v.Equals(value)
+    | (false, _) -> false
+```
+
+This code is performant, but it exposes the mutation-based data structure that callers are responsible for maintaining. This can be wrapped inside of a class with no underlying members that can change:
+
+```fsharp
+open System.Collections.Generic
+
+/// The results of computing the LALR(1) closure of an LR(0) kernel
+type Closure1Table() =
+    let t = Dictionary<Item0, HashSet<TerminalIndex>>()
+
+    member __.Add(key, value) =
+        if not (t.ContainsKey(key)) then
+            t.Add(key, value)
+        else
+            t.[key] <- value
+
+    member __.Count = t.Count
+
+    member __.Contains(key, value) =
+        match t.TryGetValue(key) with
+        | (true, v) -> v.Equals(value)
+        | (false, _) -> false
+```
+
+`Closure1Table` encapsulates the underlying mutation-based data structure, thereby not forcing callers to maintain the underlying data structure. Classes are a powerful way to encapsulate data and routines that are mutation-based without exposing the details to callers.
+
+### Prefer `let mutable` to reference cells
+
+Reference cells are a way to represent the reference to a value rather than the value itself. Although they can be used for performance-critical code, they are generally not recommended. Consider the following example:
+
+```fsharp
+let kernels =
+    let acc = ref Set.empty
+
+    processWorkList startKernels (fun kernel ->
+        if not ((!acc).Contains(kernel)) then
+            acc := (!acc).Add(kernel)
+        ...)
+
+    !acc |> Seq.toList
+```
+
+The use of a reference cell now "pollutes" all subsequent code with having to dereference and re-reference the underlying data. Instead, consider `let mutable`:
+
+```fsharp
+let kernels =
+    let mutable acc = Set.empty
+
+    processWorkList startKernels (fun kernel ->
+        if not (acc.Contains(kernel)) then
+            acc <- acc.Add(kernel)
+        ...)
+
+    acc |> Seq.toList
+```
+
+Aside from the single point of mutation in the middle of the lambda expression, all other code that touches `acc` can do so in a manner that is no different to the usage of a normal `let`-bound immutable value. This will make it easier to change over time.
+
+## Object programming
+
+F# has full support for objects and object-oriented (OO) concepts. Although many OO concepts are powerful and useful, not all of them are ideal to use. The following lists offer guidance on categories of OO features at a high level.
+
+**Consider using these features in many situations:**
+
+* Dot notation (`x.Length`)
+* Instance members
+* Implicit constructors
+* Static members
+* Indexer notation (`arr.[x]`)
+* Named and Optional arguments
+* Interfaces and interface implementations
+
+**Don't reach for these features first, but do judiciously apply them when they make sense:**
```

Comment:
I think you need to expound on "when they make sense". Without some detail on when it makes sense, this isn't very helpful guidance.